File: Runners/CSharp/CSharpHost/Services/GameService.cs
using CSharpHost.Contracts;
using CSharpHost.Models;
using System.Diagnostics;
using System.Runtime.InteropServices;

namespace CSharpHost.Services;

public class GameService(IPlayerService playerService, IGameLogger gameLogger) : IGameService
{
    private readonly IPlayerService _playerService = playerService;
    private readonly IGameLogger _gameLogger = gameLogger;

    private string? _gameId;
    private int _column;
    private int _row;
    private int _totalPacketSize;

    public void InitGame(string gameId, int column, int row)
    {
        _gameId = gameId;
        _column = column;
        _row = row;
        var mapPacketSize = (column * row) * 4;
        _totalPacketSize = mapPacketSize + 4; // map 배열크기.(int[] 배열 크기) + player position 값(4byte).

        _gameLogger.Init(gameId, column, row);
    }

    public Task LoadPlayer(int position, string filePath, [Optional] CancellationToken cancellation)
    {
        return _playerService.LoadPlayer(position, filePath, cancellation);
    }

    public string GetPlayerName(int position)
    {
        var player = _playerService.GetPlayer(position);
        return player.GetName() ?? throw new InvalidOperationException($"{position} Player name is null");
    }

    public void InitializePlayer(int position, int column, int row)
    {
        var player = _playerService.GetPlayer(position);
        player.Initialize(position, column, row);

    }

    public async Task<int> MoveNext(GameMessage message, [Optional] CancellationToken cancellation)
    {
        var (turn, position, map, current) = message;

        try
        {
            return await Task.Run(() =>
            {
                var player = _playerService.GetPlayer(position);

                var stopwatch = Stopwatch.StartNew();
                var direction = player.MoveNext(map, current);
                Debug.WriteLine($"position:{position}, direction:{direction}, elapsed:{stopwatch.ElapsedMilliseconds}");
                if (direction < 0 || direction > 3)
                {
                    throw new InvalidOperationException($"The result is out of range. result:{direction}");
                }

                _gameLogger.LogPlayerAction(new GameTurn(turn, position, map, current, (MoveDirection)direction));
                return direction;
            });
        }
        catch (Exception ex)
        {
            Debug.WriteLine(ex);
            _gameLogger.LogPlayerAction(new GameTurn(turn, position, map, current, MoveDirection.Error));
            throw;
        }
    }

    public GameSet GetCurrentGameSet()
        => new GameSet(_column, _row);

    public void CleanUp()
    {
        _gameLogger.Cleanup();

        _column = 0;
        _row = 0;
        _playerService.CleanUp();        
    }
}

File: Runners/Python/PyHost/services/game_service.py
import asyncio
import time

class GameService:
    def __init__(self, player_service):
        self._player_service = player_service
        self._column = 0
        self._row = 0
        self._total_packet_size = 0

    def init_game(self, column, row):
        self._column = column
        self._row = row
        map_packet_size = (column * row) * 4
        self._total_packet_size = map_packet_size + 4

    async def load_player(self, position, file_path, cancellation_token=None):
        await self._player_service.load_player(position, file_path, cancellation_token)

    def get_player_name(self, position):
        player = self._player_service.get_player(position)
        if player:
            return player.get_name()
        raise Exception(f"{position} Player name is null")

    def initialize_player(self, position, column, row):
        player = self._player_service.get_player(position)
        if player:
            player.initialize(position, column, row)
        else:
            raise Exception(f"Player at position {position} not found")

    async def move_next(self, message, cancellation_token=None):
        try:
            position, map_data, current = message

            return await asyncio.to_thread(self._move_next_sync, position, map_data, current)
        except Exception as ex:
            raise  Exception(f"Error during move_next: {ex}")

    def _move_next_sync(self, position, map_data, current):
        player = self._player_service.get_player(position)
        if player is None:
            raise Exception(f"Player at position {position} not found")

        start_time = time.time()
        direction = player.move_next(map_data, current)
        elapsed_time = (time.time() - start_time) * 1000
        print(f"position: {position}, direction: {direction}, elapsed: {elapsed_time} ms")

        if direction < 0 or direction > 3:
            raise Exception(f"The result is out of range. result: {direction}")

        return direction

    def get_current_game_set(self):
        return GameSet(self._column, self._row)

    def clean_up(self):
        self._column = 0
        self._row = 0
        self._player_service.clean_up()